overlord/assertstate/validation_set_tracking.go

```diff
@@ -0,0 +1,130 @@
+type ValidationSetTracking struct {
+	AccountID string            `json:"account"`
+	Name      string            `json:"name"`
+	Mode      ValidationSetMode `json:"mode"`
+
+	// PinnedAt is an optional pinned sequence point, or 0 if not pinned.
+	PinnedAt int `json:"pinned-at,omitempty"`
+
+	// Current is the current sequence point.
+	Current int `json:"cuurent,omitempty"`
+}
+
+// ValidationSetKey formats the given account id and name into a validation set key.
+func ValidationSetKey(accountID, name string) string {
+	return fmt.Sprintf("%s/%s", accountID, name)
+}
+
+// UpdateValidationSet updates ValidationSetTracking.
+// The method assumes valid tr fields.
+func UpdateValidationSet(st *state.State, tr *ValidationSetTracking) {
+	var vsmap map[string]*json.RawMessage
+	err := st.Get("validation-set-tracking", &vsmap)
+	if err != nil && err != state.ErrNoState {
+		panic("internal error: cannot unmarshal validation-set-tracking state: " + err.Error())
+	}
+	if vsmap == nil {
+		vsmap = make(map[string]*json.RawMessage)
+	}
+	data, err := json.Marshal(tr)
+	if err != nil {
+		panic("internal error: cannot marshal validation set tracking data: " + err.Error())
+	}
+	raw := json.RawMessage(data)
+	key := ValidationSetKey(tr.AccountID, tr.Name)
+	vsmap[key] = &raw
+	st.Set("validation-set-tracking", vsmap)
+}
+
+// DeleteValidationSet deletes a validation set for the given accoundID and name.
+// It is not an error to delete a non-existing one.
+func DeleteValidationSet(st *state.State, accountID, name string) error {
+	var vsmap map[string]*json.RawMessage
+	err := st.Get("validation-set-tracking", &vsmap)
+	if err != nil && err != state.ErrNoState {
+		panic("internal error: cannot unmarshal validation-set-tracking state: " + err.Error())
+	}
+	if len(vsmap) == 0 {
+		return nil
+	}
+	delete(vsmap, ValidationSetKey(accountID, name))
+	st.Set("validation-set-tracking", vsmap)
+	return nil
+}
+
+// GetValidationSet retrieves the ValidationSetTracking for the given account and name.
+func GetValidationSet(st *state.State, accountID, name string, tr *ValidationSetTracking) error {
+	if tr == nil {
+		return fmt.Errorf("internal error: tr is nil")
+	}
+
+	*tr = ValidationSetTracking{}
+
+	var vset map[string]*json.RawMessage
+	err := st.Get("validation-set-tracking", &vset)
+	if err != nil {
+		return err
+	}
+	key := ValidationSetKey(accountID, name)
+	raw, ok := vset[key]
+	if !ok {
+		return state.ErrNoState
+	}
+	err = json.Unmarshal([]byte(*raw), &tr)
+	if err != nil {
+		return fmt.Errorf("cannot unmarshal validation set tracking: %v", err)
+	}
+	return nil
+}
+
+// ValidationSets retrieves all ValidationSetTracking data.
+func ValidationSets(st *state.State) (map[string]*ValidationSetTracking, error) {
+	var vsmap map[string]*ValidationSetTracking
+	if err := st.Get("validation-set-tracking", &vsmap); err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+
+	validationSetMap := make(map[string]*ValidationSetTracking, len(vsmap))
```

Comment:
What is the purpose of this copy? Cannot you just return `vsmap`?

Reply:
Yes, totally, thanks for spotting, that was a leftover from refactor where initially this method had to iterate and mange the result from state before returning.